### PR TITLE
Merged aikido_perception and aikido_rviz into aikido

### PR DESCRIPTION
**Pull request by mkoval**
This change is [Reviewable](https://reviewable.io/reviews/personalrobotics/aikido/107)

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -185,12 +155,11 @@ if(${DOXYGEN_FOUND})
...
       "${DOXYGEN_TAGFILES} dart.tag=http://dartsim.github.io/dart/")
 
-    if(${OMPL_FOUND})
-      file(DOWNLOAD "http://ompl.kavrakilab.org/core/ompl.tag"
-        "${DOXYGEN_WORKING_DIR}/ompl.tag")
-      set(DOXYGEN_TAGFILES
-        "${DOXYGEN_TAGFILES} ompl.tag=http://ompl.kavrakilab.org/")
-    endif()
+    # TODO: Only download this if planner_ompl is being built.
+    file(DOWNLOAD "http://ompl.kavrakilab.org/core/ompl.tag"
```
Comment (psigen) [inline CMakeLists.txt]: Can we do this by checking for the `aikido_ompl` component target?

File: CMakeLists.txt [hunk 2]
```diff
@@ -206,52 +175,59 @@ if(${DOXYGEN_FOUND})
   add_custom_target(docs
     DEPENDS "${DOXYGEN_HTML_INDEX}"
     COMMENT "Generating documentation.")
+
+  # Create the "view_docs" target that opens the documentation.
+  if(APPLE)
+    set(open_command "open")
+  else()
+    set(open_command "xdg-open")
+  endif()
```
Comment (psigen) [inline CMakeLists.txt]: But what if we are building on windows? :smiling_imp:

File: CMakeLists.txt [hunk 3]
```diff
@@ -206,52 +175,59 @@ if(${DOXYGEN_FOUND})
...
-)
+  DESTINATION "${INCLUDE_INSTALL_DIR}")
 
 # Install the package.xml file (to satisfy REP-136).
 install(FILES "package.xml"
-  DESTINATION "share/${PROJECT_NAME}"
-)
+  DESTINATION "share/${PROJECT_NAME}")
+
+install_component_exports()
```
Comment (psigen) [inline CMakeLists.txt]: Why is this line here, but then commented out below?

File: cmake/aikidoConfig.cmake.in [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+    if(NOT is_visited)
+      set(component_include
+        "${CMAKE_CURRENT_LIST_DIR}/aikido_${component}Component.cmake")
+      set(target_include
+        "${CMAKE_CURRENT_LIST_DIR}/aikido_${component}Targets.cmake")
+
+      if(EXISTS "${component_include}" AND EXISTS "${target_include}")
+        include("${component_include}")
+
+        set_property(DIRECTORY PROPERTY "aikido_${component}_FOUND" TRUE)
```
Comment (psigen) [inline cmake/aikidoConfig.cmake.in]: Why do we use `DIRECTORY` scope on these properties? It seems like `find_package` would normally have global scope, but I'm sure there's some bizarre convention at work here?

File: cmake/aikidoConfig.cmake.in [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+        # This is not necessary in recent versions of cmake, which supports
+        # transitive includes via INTERFACE_INCLUDE_DIRECTORIES.
+        foreach(dependency ${aikido_${component}_DEPENDENCIES})
+          set_property(DIRECTORY APPEND
+            PROPERTY "aikido_${component}_INCLUDE_DIRS"
+            ${aikido_${dependency}_INCLUDE_DIRS})
+          set_property(DIRECTORY APPEND
+            PROPERTY "aikido_${component}_DEFINITIONS"
+            ${aikido_${dependency}_DEFINITIONS})
+          # LIBRARIES is automatically inherited via INTERFACE_LINK_LIBRARIES.
```
Comment (psigen) [inline cmake/aikidoConfig.cmake.in]: Would it make sense to do a CMake version check on avoiding this entire loop in versions that support `INTERFACE_XXXX` properties? (and export the correspondingly named properties)

File: cmake/aikidoConfig.cmake.in [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+
+  aikido_export_variable(aikido_INCLUDE_DIRS)
+  aikido_export_variable(aikido_DEFINITIONS)
+  aikido_export_variable(aikido_LIBRARIES)
+endfunction()
+
+#==============================================================================
+
+if(aikido_FIND_COMPONENTS)
+  # TODO: Should we append this to aikido_<C>_INCLUDE_DIRS above?
```
Comment (psigen) [inline cmake/aikidoConfig.cmake.in]: Do we still need this TODO? Looks like we are doing this now.

File: cmake/Components.cmake [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+  BRIEF_DOCS "Global include directories used by all components."
+  FULL_DOCS "Global include directories used by all components."
+)
+define_property(GLOBAL PROPERTY "${global_prefix}COMPONENTS"
+  BRIEF_DOCS "List all known Aikido components."
+  FULL_DOCS "List all known Aikido components."
+)
+
+#==============================================================================
+function(add_component component)
```
Comment (psigen) [inline cmake/Components.cmake]: Can we turn this into a metafunction using [CMakeParseArguments](https://cmake.org/cmake/help/v3.0/module/CMakeParseArguments.html) so that we have single declarations that look like:
```cmake
add_component(
    NAME control
    TARGETS "${PROJECT_NAME}_control"
    DEPENDENCIES
        control
        statespace
        trajectory
)
```

Comment (psigen) [inline cmake/Components.cmake]: Also, is there a need to make this `add_aikido_component()`?

File: CMakeLists.txt (hunk 2 above)
Comment (mkoval) [inline CMakeLists.txt]: You're welcome to send a pull request for that case. :innocent:

File: cmake/aikidoConfig.cmake.in (hunk 4 above)
Comment (mkoval) [inline cmake/aikidoConfig.cmake.in]: These are intended to be parent-scoped output variables. I am storing them in `DIRECTORY` properties because:
- `PARENT_SCOPE` only works for one level of scope
- using a `macro` clobbers all local variables (e.g. `component`) during recursion
- a `GLOBAL` property will pollute strictly more scope

The ideal solution would be to use a `PARENT_SCOPE` variable and _very carefully_ forward all of the variables recursively up the recursion stack.

Another option is to use `GLOBAL` or `CACHE`d properties, but prefix them to make it clear that they're an implementation detail not intended for external use (e.g. `aikido_internal_`). Otherwise, setting these variables defeats the purpose of carefully setting `CACHE`d properties for _only the components that the user asks for_.

In any case: this is an implementation detail. I'd prefer to change this in another pull request.

File: cmake/aikidoConfig.cmake.in (hunk 5 above)
Comment (mkoval) [inline cmake/aikidoConfig.cmake.in]: I am open to cutting `_INCLUDE_DIRS` and `_DEFINITIONS` entirely. Aikido depends on CMake 2.8.12, which already has support for those `INTERFACE_` properties.

File: cmake/aikidoConfig.cmake.in (hunk 6 above)
Comment (mkoval) [inline cmake/aikidoConfig.cmake.in]: Removed.

File: cmake/Components.cmake (hunk 7 above)
Comment (mkoval) [inline cmake/Components.cmake]: > Can we turn this into a metafunction using CMakeParseArguments so that we have single declarations that look like: <snip>

I am open to adding syntactic sugar around `add_component`, but I would prefer to do that in a future pull request.

> Also, is there a need to make this `add_aikido_component()`?

I am not sure what to do here. I intentionally _did not_ add an `aikido` prefix to these variables because I would like to reuse the same `Components.cmake` file in other projects (e.g. DART). However, this does create the potential for name collisions.

Is there any way to namespace a `.cmake` file? I attempted to define the functions like this:
```cmake
function(${prefix}_add_component)
  # ...
endfunction()
```
Unfortunately, I couldn't figure out how to _call_ that function from this file. These did *not* work:
```cmake
${prefix}_add_component() # nope
"${prefix}_add_component"() # nope
```

File: CMakeLists.txt (hunk 3 above)
Comment (mkoval) [inline CMakeLists.txt]: Fixed.

File: CMakeLists.txt (hunk 1 above)
Comment (mkoval) [inline CMakeLists.txt]: Done.

**Comment by mkoval** [comment]
> Couldn't we just always do both define_ and set_property?
> ```cmake
define_property(GLOBAL PROPERTY "aikido_${component}_FOUND")
set_property(GLOBAL PROPERTY "aikido_${component}_FOUND" FALSE)
```

We _could_ do this. It's a bit funny because `define_property` operates on the _target type_ of `DIRECTORY`, not on a _specific directory_. This means that we're defining these internal properties on _every directory_ that CMake encounters, even if they are not set.

I prefer keeping it as it is until we replace these internal properties with something more elegant (see above).

File: cmake/Components.cmake (hunk 7 above)
Comment (psigen) [inline cmake/Components.cmake]: Well, I don't think we should necessarily assume that we are making components for something named `PROJECT_NAME`.

I think one reasonable way to do this would be to actually add another property to these functions like:
```cmake
function(add_component package_name component)
    [...]
endfunction()

macro(setup_component_helpers package_name)
    function(add_${package_name}_component)
        add_component(PACKAGE aikido ${ARGV}})
    endfunction()
    [...]
endmacro()
```
and then in _aikido_'s top-level cmake do something like:
```cmake
include(Components.cmake)
setup_component_helpers(aikido)

add_aikido_component(...)
```
(based on http://stackoverflow.com/a/31947751)

File: cmake/aikidoConfig.cmake.in (hunk 5 above)
Comment (psigen) [inline cmake/aikidoConfig.cmake.in]: I think this would be pretty great.  It would eliminate this loop entirely and be forward compliant with transitioning to just using the builtin transitive dependency system in later CMake versions.

File: cmake/aikidoConfig.cmake.in (hunk 4 above)
Comment (psigen) [inline cmake/aikidoConfig.cmake.in]: Ok, I see.

**Comment by psigen** [comment]
Cool, can we file issues for the syntactic sugar on `add_component` and cleaning up the `DIRECTORY` properties with `PARENT_SCOPE` propagation?

File: cmake/aikidoConfig.cmake.in (hunk 5 above)
Comment (mkoval) [inline cmake/aikidoConfig.cmake.in]: I *think* we already have the builtin transitive dependency system as of 2.8.12!

File: cmake/Components.cmake [hunk 8]
```diff
@@ -0,0 +1,145 @@
...
+function(install_component_exports)
+  get_property(components GLOBAL PROPERTY "${global_prefix}COMPONENTS")
+
+  set(output_prefix "${CMAKE_CURRENT_BINARY_DIR}/${CONFIG_INSTALL_DIR}")
+
+  foreach(component ${components})
+    set(target "${component_prefix}${component}")
+
+    # TODO: Replace this manual generation with a configure_file.
+    set(output_path
```
Comment (ClintLiddick) [inline cmake/Components.cmake]: This is the only part that feels not great. I'm sure it works, but writing directly to a file from CMake seems like an abuse. What would the config file look like that you can reuse multiple times here?

File: cmake/Components.cmake (hunk 7 above)
Comment (ClintLiddick) [inline cmake/Components.cmake]: That sounds like a good solution. Everything else looks good to me, so I'll make this change and then it gets my :+1:

File: cmake/Components.cmake (hunk 8 above)
Comment (psigen) [inline cmake/Components.cmake]: I'm not sure I follow the issue here: CMake is writing a CMake configuration, not something that's part of the build.  Is there some part of the build system other than CMake that is qualified to write CMake configuration files?

Comment (psigen) [inline cmake/Components.cmake]: Ohh, you mean the manual appending?  Yeah, as the `TODO` states, the configure_file solution would be better here, but it is kind of pain to implement.

File: cmake/Components.cmake (hunk 7 above)
Comment (ClintLiddick) [inline cmake/Components.cmake]: Turns out this is a mess of mixed macro and argument substitution.... I'm trying to make it not horrible now.

Comment (ClintLiddick) [inline cmake/Components.cmake]: Ok, I am officially giving up on namespacing. If you use a `macro` to generate the namespaced functions you trash all usage of `ARGN`, etc making variadic functions impossible. If you use a `function` then text substitution stops as the inner `function` definitions, and you cannot pass `${package_name}` to the aliased function (because that code isn't executed until the inner function is called, outside the scope of the defining function).

The question is do we want to wrap everything in a setup macro and do all the text substitution instead of just using `${PROJECT_NAME}`? @mkoval thoughts?

**Comment by ClintLiddick** [comment]
I refactored the `Components.cmake` to namespace all the functions and use a function to initialize the global properties. I'm GTG.
